File: typescript/samples/agents/basic-m365/src/TestingChatClient.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

/**
 * Testing wrapper for OpenAI client that supports recording and playback of LLM interactions.
 * This enables deterministic testing by recording real LLM responses and playing them back.
 */

import OpenAI from 'openai'
import * as fs from 'fs'
import * as path from 'path'
import * as crypto from 'crypto'

export class TestingChatClient {
  private readonly realClient: OpenAI | null
  private readonly recordingsDir: string
  private readonly modelId: string
  private readonly recordMode: boolean
  private readonly playbackMode: boolean
  private callCount: number = 0

  constructor(
    realClient: OpenAI | null,
    recordingsDir: string,
    modelId: string,
    recordMode: boolean = false,
    playbackMode: boolean = false
  ) {
    if (recordMode && playbackMode) {
      throw new Error('Cannot enable both record and playback mode simultaneously')
    }

    if (recordMode && !realClient) {
      throw new Error('Real client required for recording mode')
    }

    if (playbackMode && !fs.existsSync(recordingsDir)) {
      throw new Error(`Recordings directory not found: ${recordingsDir}`)
    }

    this.realClient = realClient
    this.recordingsDir = recordingsDir
    this.modelId = modelId
    this.recordMode = recordMode
    this.playbackMode = playbackMode

    if (this.recordMode) {
      fs.mkdirSync(this.recordingsDir, { recursive: true })
      console.log(`📹 LLM Recording enabled: ${this.recordingsDir}`)
    } else if (this.playbackMode) {
      console.log(`▶️  LLM Playback enabled: ${this.recordingsDir}`)
      console.log('   Using recorded LLM responses (test mode)')
    }
  }

  async createCompletion(
    messages: OpenAI.Chat.ChatCompletionMessageParam[],
    tools?: OpenAI.Chat.ChatCompletionTool[]
  ): Promise<OpenAI.Chat.ChatCompletion> {
    this.callCount++
    const callId = this.callCount
    const hashKey = this.computeRequestHash(messages, tools || [])

    if (this.playbackMode) {
      return this.playbackResponse(callId, hashKey, messages, tools || [])
    }

    // Call real LLM (works in both normal and recording mode)
    if (!this.realClient) {
      throw new Error('Real client not available. Use recording mode with a valid OpenAI client.')
    }

    const response = await this.realClient.chat.completions.create({
      model: this.modelId,
      messages: messages,
      tools: tools && tools.length > 0 ? tools : undefined
    })

    if (this.recordMode) {
      await this.recordInteraction(callId, hashKey, messages, tools || [], response)
    }

    return response
  }

  private playbackResponse(
    callId: number,
    hashKey: string,
    messages: OpenAI.Chat.ChatCompletionMessageParam[],
    tools: OpenAI.Chat.ChatCompletionTool[]
  ): OpenAI.Chat.ChatCompletion {
    const responseFile = path.join(this.recordingsDir, `${hashKey}.response.json`)

    if (!fs.existsSync(responseFile)) {
      throw new Error(
        `No recorded LLM response found for request hash: ${hashKey}\n` +
        `Expected file: ${responseFile}\n\n` +
        `This usually means:\n` +
        `1. Tests need to be run in generation mode first: RECORD_LLM=true\n` +
        `2. The request parameters have changed (different hash)\n` +
        `3. The recording file was deleted\n\n` +
        `Request details:\n` +
        `  Messages: ${messages.length} messages\n` +
        `  Tools: ${tools.length > 0 ? 'provided' : 'null'}\n`
      )
    }

    console.log(`  ▶️  Replaying LLM call #${callId}: ${hashKey}`)

    const recordingData = JSON.parse(fs.readFileSync(responseFile, 'utf-8'))
    const responseData = recordingData.response

    return this.deserializeCompletion(responseData)
  }

  private async recordInteraction(
    callId: number,
    hashKey: string,
    messages: OpenAI.Chat.ChatCompletionMessageParam[],
    tools: OpenAI.Chat.ChatCompletionTool[],
    response: OpenAI.Chat.ChatCompletion
  ): Promise<void> {
    console.log(`  📹 Recording LLM call #${callId}: ${hashKey}`)

    // Save request
    const requestFile = path.join(this.recordingsDir, `${hashKey}.request.json`)
    const requestData = {
      callId,
      timestamp: new Date().toISOString(),
      hash: hashKey,
      model: this.modelId,
      messages: this.normalizeMessages(messages),
      tools: tools.length > 0 ? tools.map(t => ({
        type: 'function',
        function: {
          name: t.function.name,
          description: t.function.description || '',
          parameters: t.function.parameters || {}
        }
      })) : undefined
    }

    fs.writeFileSync(requestFile, JSON.stringify(requestData, null, 2))

    // Save response
    const responseFile = path.join(this.recordingsDir, `${hashKey}.response.json`)
    const choice = response.choices[0]

    // Extract content
    const contentList: any[] = []
    if (choice.message.content) {
      contentList.push({ text: choice.message.content })
    }

    // Extract tool calls
    const toolCallsList: any[] = []
    if (choice.message.tool_calls) {
      for (const tc of choice.message.tool_calls) {
        toolCallsList.push({
          id: tc.id,
          type: 'Function',
          function: {
            name: tc.function.name,
            arguments: tc.function.arguments
          }
        })
      }
    }

    const responseData = {
      callId,
      timestamp: new Date().toISOString(),
      hash: hashKey,
      response: {
        id: response.id,
        model: response.model,
        created: response.created,
        finishReason: choice.finish_reason ? this.capitalizeFirstLetter(choice.finish_reason) : 'Stop',
        content: contentList,
        toolCalls: toolCallsList
      }
    }

    fs.writeFileSync(responseFile, JSON.stringify(responseData, null, 2))
  }

  private computeRequestHash(
    messages: OpenAI.Chat.ChatCompletionMessageParam[],
    tools: OpenAI.Chat.ChatCompletionTool[]
  ): string {
    // Build request dict matching C# format exactly
    const requestDict: any = {
      model: this.modelId,
      messages: this.normalizeMessages(messages),
      temperature: 0.0
    }

    if (tools && tools.length > 0) {
      requestDict.tools = tools.map(t => ({
        type: 'function',
        function: {
          name: t.function.name,
          description: t.function.description || '',
          parameters: JSON.stringify(t.function.parameters || {})
        }
      }))
    }

    // Serialize to JSON with sorted keys (recursively)
    const sortedDict = this.sortJsonKeys(requestDict)
    const jsonStr = JSON.stringify(sortedDict)

    // Log the JSON being hashed for debugging
    console.log(`🔍 [TypeScript] Computing hash for:`)
    console.log(`   JSON length: ${jsonStr.length} chars`)
    console.log(`   FULL JSON: ${jsonStr}`)
    fs.writeFileSync('/tmp/typescript_json.txt', jsonStr)

    // Hash and truncate (SHA256, first 16 chars)
    const hash = crypto.createHash('sha256').update(jsonStr, 'utf8').digest('hex')
    const hashStr = hash.substring(0, 16).toLowerCase()
    console.log(`   Hash: ${hashStr}`)
    return hashStr
  }

  private normalizeMessages(messages: OpenAI.Chat.ChatCompletionMessageParam[]): any[] {
    return messages.map((msg: any) => {
      const normalized: any = { role: msg.role }

      // Add content if present
      if (msg.content !== undefined && msg.content !== null) {
        normalized.content = msg.content
      }

      // Add tool_calls if present
      if (msg.tool_calls && msg.tool_calls.length > 0) {
        normalized.tool_calls = msg.tool_calls.map((tc: any) => ({
          id: tc.id,
          type: tc.type || 'function',
          function: {
            name: tc.function.name,
            arguments: tc.function.arguments
          }
        }))
      }

      // Add tool_call_id if present
      if (msg.tool_call_id) {
        normalized.tool_call_id = msg.tool_call_id
      }

      return normalized
    })
  }

  private deserializeCompletion(responseElement: any): OpenAI.Chat.ChatCompletion {
    const completionId = responseElement.id || 'mock-completion'
    const model = responseElement.model || 'unknown'
    const finishReasonStr = responseElement.finishReason || 'Stop'

    // Map finish reason
    const finishReasonMap: Record<string, string> = {
      'ToolCalls': 'tool_calls',
      'Stop': 'stop',
      'Length': 'length'
    }
    const finishReason = finishReasonMap[finishReasonStr] || 'stop'

    // Parse content
    let content: string | null = null
    const contentArray = responseElement.content || []
    if (contentArray.length > 0) {
      content = contentArray[0].text || ''
    }

    // Parse tool calls
    let toolCalls: OpenAI.Chat.ChatCompletionMessageToolCall[] | undefined = undefined
    const toolCallsArray = responseElement.toolCalls || []
    if (toolCallsArray.length > 0) {
      toolCalls = toolCallsArray.map((tc: any) => ({
        id: tc.id,
        type: 'function' as const,
        function: {
          name: tc.function.name,
          arguments: tc.function.arguments
        }
      }))
    }

    // Build mock completion object
    const completion: OpenAI.Chat.ChatCompletion = {
      id: completionId,
      object: 'chat.completion',
      created: Math.floor(Date.now() / 1000),
      model: model,
      choices: [{
        index: 0,
        message: {
          role: 'assistant',
          content: content,
          refusal: null,
          tool_calls: toolCalls
        },
        finish_reason: finishReason as any,
        logprobs: null
      }],
      usage: {
        prompt_tokens: 0,
        completion_tokens: 0,
        total_tokens: 0
      }
    }

    return completion
  }

  /**
   * Recursively sorts all keys in an object or array to ensure consistent JSON serialization.
   * This matches the behavior of C#'s SortJsonKeys method.
   */
  private sortJsonKeys(value: any): any {
    if (value === null || value === undefined) {
      return value
    }

    // Handle arrays by recursively sorting each element
    if (Array.isArray(value)) {
      return value.map(item => this.sortJsonKeys(item))
    }

    // Handle objects by sorting keys and recursively sorting values
    if (typeof value === 'object') {
      const sorted: any = {}
      const keys = Object.keys(value).sort()
      for (const key of keys) {
        sorted[key] = this.sortJsonKeys(value[key])
      }
      return sorted
    }

    // Primitive values (string, number, boolean) pass through
    return value
  }

  private capitalizeFirstLetter(str: string): string {
    if (!str) return str
    // Handle snake_case to PascalCase (tool_calls -> ToolCalls)
    return str.split('_').map(word =>
      word.charAt(0).toUpperCase() + word.slice(1)
    ).join('')
  }
}

File: python/samples/agents/emoji-chat/src/testing_chat_client.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

"""
Testing wrapper for OpenAI AsyncOpenAI that supports recording and playback of LLM interactions.
This enables deterministic testing by recording real LLM responses and playing them back.
"""

import hashlib
import json
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional


class TestingChatClient:
    """Testing wrapper for AsyncOpenAI that supports recording and playback of LLM interactions."""

    def __init__(
        self,
        real_client: Optional[Any],
        recordings_dir: str,
        model_id: str,
        record_mode: bool = False,
        playback_mode: bool = False
    ):
        """Initialize TestingChatClient.

        Args:
            real_client: Real AsyncOpenAI client (required for normal and recording mode)
            recordings_dir: Directory for storing/loading recordings
            model_id: Model identifier
            record_mode: If True, records LLM interactions to disk
            playback_mode: If True, replays recorded LLM interactions

        Raises:
            ValueError: If both record_mode and playback_mode are enabled
            ValueError: If record_mode is enabled without a real_client
            FileNotFoundError: If playback_mode is enabled but recordings_dir doesn't exist
        """
        if record_mode and playback_mode:
            raise ValueError("Cannot enable both record and playback mode simultaneously")

        if record_mode and real_client is None:
            raise ValueError("Real client required for recording mode")

        recordings_path = Path(recordings_dir)
        if playback_mode and not recordings_path.exists():
            raise FileNotFoundError(f"Recordings directory not found: {recordings_dir}")

        self._real_client = real_client
        self._recordings_dir = recordings_path
        self._model_id = model_id
        self._record_mode = record_mode
        self._playback_mode = playback_mode
        self._call_count = 0

        if self._record_mode:
            self._recordings_dir.mkdir(parents=True, exist_ok=True)
            print(f"📹 LLM Recording enabled: {self._recordings_dir}")
        elif self._playback_mode:
            print(f"▶️  LLM Playback enabled: {self._recordings_dir}")
            print("   Using recorded LLM responses (test mode)")

    async def create_completion(
        self,
        messages: List[Dict[str, Any]],
        tools: Optional[List[Dict[str, Any]]] = None,
        **kwargs
    ):
        """Create a chat completion, either by calling the real LLM or replaying a recording.

        Args:
            messages: List of conversation messages
            tools: Optional list of tool definitions
            **kwargs: Additional arguments (ignored for hashing compatibility)

        Returns:
            Chat completion response (either real or replayed)
        """
        self._call_count += 1
        call_id = self._call_count
        hash_key = self._compute_request_hash(messages, tools)

        if self._playback_mode:
            return await self._playback_response(call_id, hash_key, messages, tools)

        # Call real LLM (works in both normal and recording mode)
        if self._real_client is None:
            raise ValueError("Real client not available. Use recording mode with a valid AsyncOpenAI client.")

        response = await self._real_client.chat.completions.create(
            model=self._model_id,
            messages=messages,
            tools=tools if tools else None,
            **kwargs
        )

        if self._record_mode:
            await self._record_interaction(call_id, hash_key, messages, tools, response)

        return response

    async def _playback_response(
        self,
        call_id: int,
        hash_key: str,
        messages: List[Dict[str, Any]],
        tools: Optional[List[Dict[str, Any]]]
    ):
        """Replay a recorded LLM response.

        Args:
            call_id: Sequential call number
            hash_key: Hash of the request
            messages: Request messages
            tools: Request tools

        Returns:
            Mock completion object matching the recorded response

        Raises:
            FileNotFoundError: If no recording exists for this hash
        """
        response_file = self._recordings_dir / f"{hash_key}.response.json"

        if not response_file.exists():
            raise FileNotFoundError(
                f"No recorded LLM response found for request hash: {hash_key}\n"
                f"Expected file: {response_file}\n\n"
                f"This usually means:\n"
                f"1. Tests need to be run in generation mode first: RECORD_LLM=true\n"
                f"2. The request parameters have changed (different hash)\n"
                f"3. The recording file was deleted\n\n"
                f"Request details:\n"
                f"  Messages: {len(messages)} messages\n"
                f"  Tools: {'provided' if tools else 'null'}\n"
            )

        print(f"  ▶️  Replaying LLM call #{call_id}: {hash_key}")

        with open(response_file, 'r') as f:
            recording = json.load(f)

        response_data = recording["response"]
        return self._deserialize_completion(response_data)

    async def _record_interaction(
        self,
        call_id: int,
        hash_key: str,
        messages: List[Dict[str, Any]],
        tools: Optional[List[Dict[str, Any]]],
        response: Any
    ):
        """Record an LLM interaction to disk.

        Args:
            call_id: Sequential call number
            hash_key: Hash of the request
            messages: Request messages
            tools: Request tools
            response: LLM response to record
        """
        print(f"  📹 Recording LLM call #{call_id}: {hash_key}")

        # Save request
        request_file = self._recordings_dir / f"{hash_key}.request.json"
        request_data = {
            "callId": call_id,
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "hash": hash_key,
            "model": self._model_id,
            "messages": self._normalize_messages(messages),
            "tools": [
                {
                    "type": "function",
                    "function": {
                        "name": t["function"]["name"],
                        "description": t["function"].get("description", ""),
                        "parameters": t["function"].get("parameters", {})
                    }
                }
                for t in (tools or [])
            ] if tools else None
        }

        with open(request_file, 'w') as f:
            json.dump(request_data, f, indent=2)

        # Save response
        response_file = self._recordings_dir / f"{hash_key}.response.json"
        choice = response.choices[0]

        # Extract content
        content_list = []
        if choice.message.content:
            content_list.append({"text": choice.message.content})

        # Extract tool calls
        tool_calls_list = []
        if choice.message.tool_calls:
            for tc in choice.message.tool_calls:
                tool_calls_list.append({
                    "id": tc.id,
                    "type": "Function",
                    "function": {
                        "name": tc.function.name,
                        "arguments": tc.function.arguments
                    }
                })

        response_data = {
            "callId": call_id,
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "hash": hash_key,
            "response": {
                "id": response.id,
                "model": response.model,
                "created": response.created,
                "finishReason": choice.finish_reason.title() if choice.finish_reason else "Stop",
                "content": content_list,
                "toolCalls": tool_calls_list
            }
        }

        with open(response_file, 'w') as f:
            json.dump(response_data, f, indent=2)

    def _compute_request_hash(
        self,
        messages: List[Dict[str, Any]],
        tools: Optional[List[Dict[str, Any]]]
    ) -> str:
        """Compute SHA256 hash of request parameters.

        Uses exact same algorithm as C# TestingChatClient for compatibility.

        Args:
            messages: Conversation messages
            tools: Tool definitions

        Returns:
            First 16 characters of SHA256 hash (lowercase hex)
        """
        # Build request dict matching C# format exactly
        request_dict = {
            "model": self._model_id,
            "messages": self._normalize_messages(messages),
            "temperature": 0.0
        }

        if tools:
            request_dict["tools"] = [
                {
                    "type": "function",
                    "function": {
                        "name": t["function"]["name"],
                        "description": t["function"].get("description", ""),
                        "parameters": json.dumps(t["function"].get("parameters", {}), separators=(',', ':'))
                    }
                }
                for t in tools
            ]

        # Serialize to JSON with sorted keys
        # IMPORTANT: ensure_ascii=False keeps Unicode characters (emojis, arrows) unescaped
        # This matches C# JSON serialization for consistent cross-language hashing
        json_str = json.dumps(request_dict, sort_keys=True, separators=(',', ':'), ensure_ascii=False)

        # Log the JSON being hashed for debugging
        print(f"🔍 [Python] Computing hash for:")
        print(f"   JSON length: {len(json_str)} chars")
        print(f"   FULL JSON: {json_str}")
        with open('/tmp/python_json.txt', 'w') as f:
            f.write(json_str)

        # Hash and truncate (SHA256, first 16 chars)
        hash_obj = hashlib.sha256(json_str.encode('utf-8'))
        hash_str = hash_obj.hexdigest()[:16].lower()
        print(f"   Hash: {hash_str}")
        return hash_str

    def _normalize_messages(self, messages: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Normalize messages for consistent hashing.

        Args:
            messages: Raw message list

        Returns:
            Normalized message list matching C# format
        """
        normalized = []
        for msg in messages:
            norm_msg = {"role": msg["role"]}

            # Add content if present
            if "content" in msg and msg["content"] is not None:
                norm_msg["content"] = msg["content"]

            # Add tool_calls if present
            if "tool_calls" in msg and msg["tool_calls"]:
                norm_msg["tool_calls"] = [
                    {
                        "id": tc["id"],
                        "type": tc.get("type", "function"),
                        "function": {
                            "name": tc["function"]["name"],
                            "arguments": tc["function"]["arguments"]
                        }
                    }
                    for tc in msg["tool_calls"]
                ]

            # Add tool_call_id if present
            if "tool_call_id" in msg:
                norm_msg["tool_call_id"] = msg["tool_call_id"]

            normalized.append(norm_msg)

        return normalized

    def _deserialize_completion(self, response_element: Dict[str, Any]):
        """Deserialize recorded completion data into a mock completion object.

        Args:
            response_element: Recorded response data

        Returns:
            Mock completion object compatible with OpenAI response format
        """
        completion_id = response_element.get("id", "mock-completion")
        model = response_element.get("model", "unknown")
        finish_reason_str = response_element.get("finishReason", "Stop")

        # Map finish reason
        finish_reason_map = {
            "ToolCalls": "tool_calls",
            "Stop": "stop",
            "Length": "length"
        }
        finish_reason = finish_reason_map.get(finish_reason_str, "stop")

        # Parse content
        content = None
        content_array = response_element.get("content", [])
        if content_array and len(content_array) > 0:
            content = content_array[0].get("text", "")

        # Parse tool calls
        tool_calls = None
        tool_calls_array = response_element.get("toolCalls", [])
        if tool_calls_array:
            class MockFunction:
                def __init__(self, name: str, arguments: str):
                    self.name = name
                    self.arguments = arguments

            class MockToolCall:
                def __init__(self, tc_data: Dict[str, Any]):
                    self.id = tc_data["id"]
                    self.type = tc_data.get("type", "function").lower()
                    func_data = tc_data["function"]
                    self.function = MockFunction(
                        func_data["name"],
                        func_data["arguments"]
                    )

            tool_calls = [MockToolCall(tc) for tc in tool_calls_array]

        # Build mock completion object
        class MockMessage:
            def __init__(self, content, tool_calls):
                self.content = content
                self.tool_calls = tool_calls
                self.role = "assistant"

        class MockChoice:
            def __init__(self, finish_reason, message):
                self.finish_reason = finish_reason
                self.message = message
                self.index = 0

        class MockCompletion:
            def __init__(self, completion_id, model, choices):
                self.id = completion_id
                self.model = model
                self.choices = choices
                self.created = int(datetime.utcnow().timestamp())

        message = MockMessage(content, tool_calls)
        choice = MockChoice(finish_reason, message)
        return MockCompletion(completion_id, model, [choice])
